test/tools/uri_spec_runner.ts

```diff
@@ -0,0 +1,304 @@
+    'uri',
+    'valid',
+    'warning',
+    'hosts',
+    'auth',
+    'options',
+    'credential'
+  ];
+  expect(knownTestKeys).to.include.members(Object.keys(test));
+
+  const errorMessage = `"${test.uri}"`;
+
+  const valid = test.valid && (!(test as UriTest).warning || shouldNotThrowOnWarn);
+
+  if (!valid) {
+    try {
+      new MongoClient(test.uri);
+      expect.fail(`Expected "${test.uri}" to be invalid${test.valid ? ' because of warning' : ''}`);
+    } catch (err) {
+      if (err instanceof TypeError) {
+        expect(err).to.have.property('code').equal('ERR_INVALID_URL');
+      } else if (
+        // most of our validation is MongoParseError, which does not extend from MongoAPIError
+        !(err instanceof MongoParseError) &&
+        // the rest of our validation extends from MongoAPIError
+        !(err instanceof MongoAPIError) &&
+        // mongodb-connection-string-url does not export its MongoParseError so we can't check for it directly
+        err.name !== 'MongoParseError'
+      ) {
+        throw err;
+      }
+    }
+    return;
+  }
+
+  const client = new MongoClient(test.uri);
+  const options = client.options;
+  expect(options, errorMessage).to.be.an('object').that.is.not.empty;
+
+  if ('hosts' in test && test.hosts != null) {
+    for (const [index, { host, port }] of test.hosts.entries()) {
+      const actualHost = options.hosts[index];
+      if (actualHost.host == null && actualHost.socketPath == null) {
+        expect.fail('Expected host to define "host" or "socketPath" properties');
+      }
+      if (actualHost.host != null) {
+        expect(actualHost, errorMessage).property('host').to.equal(host);
+      } else {
+        expect(actualHost, errorMessage).property('socketPath').to.equal(host);
+      }
+
+      if (port != null) expect(actualHost).property('port').to.equal(port);
+    }
+  }
+
+  // depending on whether this is a UriTest or an AuthTest, credential option values are
+  // defined in test.credential or test.auth, respectively, and additional properties
+  // to test are defined in test.options or mixed into test.credential, respectively
+  let credentialsToTest: { source?: string; password?: string; username?: string } = {};
+  let optionsToTest: Record<string, any> = 'options' in test ? test.options || {} : {};
+
+  if (!('options' in test) && test.credential != null) {
+    // handle AuthTest credential testing
+    const credentialOptions = [
+      'username',
+      'password',
+      'source',
+      'mechanism',
+      'mechanism_properties'
+    ];
+    expect(test).property('credential').to.have.all.keys(credentialOptions);
+
+    optionsToTest = (({ mechanism, mechanism_properties }) => ({
+      mechanism,
+      mechanism_properties
+    }))(test.credential);
+
+    credentialsToTest = (({ username, password, source }) => ({
+      username,
+      password,
+      source
+    }))(test.credential);
+  } else if ('auth' in test && test.auth !== null) {
```

Comment:
A type assertion would definitely help clean this up - because it's confusing that looking for `auth` in the test object is actually indicative of a UriTest, not an AuthTest 😅

Reply:
fair point, both this check and the immediately invoked function were implemented this way for expediency (i.e., the expediency of getting the hundred or so tests to be correctly processed)